### PR TITLE
[BUG] Method `TSDataset.tsdataset_idx_slice` loses hierarchical structure

### LINKED ISSUE #439: [BUG] Method `TSDataset.tsdataset_idx_slice` loses hierarchical structure
### 🐛 Bug Report

Method `TSDataset.tsdataset_idx_slice` loses hierarchical structure.

### Expected behavior

It shouldn't

### How To Reproduce

```python
from etna.datasets import TSDataset
from etna.datasets import generate_hierarchical_df


def main():
    df = generate_hierarchical_df(periods=100, n_segments=[2, 4], start_time="2021-01-01",)
    df, hierarchical_structure = TSDataset.to_hierarchical_dataset(df=df, level_columns=["level_0", "level_1"])
    ts = TSDataset(df=df, freq="D", hierarchical_structure=hierarchical_structure)

    new_ts = ts.tsdataset_idx_slice(start_idx=10, end_idx=50)
    assert new_ts.hierarchical_structure is not None


if __name__ == "__main__":
    main()
```

### Environment

_No response_

### Additional context

_No response_

### Checklist

- [X] Bug appears at the latest library version

### PR DESCRIPTION
**Pull request by brsnw250**
## Before submitting (must do checklist)

- [x] Did you read the [contribution guide](https://github.com/etna-team/etna/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs? We use Numpy format for all the methods and classes.
- [x] Did you write any new necessary tests?
- [x] Did you update the [CHANGELOG](https://github.com/etna-team/etna/blob/master/CHANGELOG.md)?

## Closing issues

closes #439

### CODE REVIEW

File: etna/datasets/tsdataset.py [hunk 1]
```diff
@@ -520,16 +520,11 @@ def tsdataset_idx_slice(self, start_idx: Optional[int] = None, end_idx: Optional
...
-        tsdataset_slice = TSDataset(df=df_slice, freq=self.freq)
-        # can't put known_future into constructor, _check_known_future fails with df_exog=None
-        tsdataset_slice.known_future = deepcopy(self.known_future)
-        tsdataset_slice._regressors = deepcopy(self.regressors)
-        if self.df_exog is not None:
-            tsdataset_slice.df_exog = self.df_exog.copy(deep=True)
-        tsdataset_slice._target_components_names = deepcopy(self._target_components_names)
-        tsdataset_slice._prediction_intervals_names = deepcopy(self._prediction_intervals_names)
-        return tsdataset_slice
+        ts_slice = deepcopy(self)
```
Comment (d-a-bunin) [inline etna/datasets/tsdataset.py:523]: Don't we want to make it in a more optimized way like in `train_test_split`?

File: etna/datasets/tsdataset.py [hunk 2]
```diff
@@ -520,16 +520,11 @@ def tsdataset_idx_slice(self, start_idx: Optional[int] = None, end_idx: Optional
...
+        ts_slice.df = ts_slice.df.iloc[start_idx:end_idx].copy(deep=None)
```
[the remark this answers is not in the document]
Comment (brsnw250) [reply etna/datasets/tsdataset.py:524]: Inspired by this [one](https://github.com/pandas-dev/pandas/blob/0691c5cf90477d3503834d983f69350f250a6ff7/pandas/core/generic.py#L7361)

File: CHANGELOG.md [hunk 3]
```diff
@@ -46,7 +46,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - **Breaking:** Bump minimum `optuna` version to 4.0 ([#599](https://github.com/etna-team/etna/pull/599))
 - **Breaking:** Bump minimum `statsforecast` version to 2.0 ([#599](https://github.com/etna-team/etna/pull/599))
 - Optimize performance of exogenous variables addition to the dataset ([#596](https://github.com/etna-team/etna/pull/596))
--
+- Update `TSDataset.tsdataset_idx_slice` method ([#618](https://github.com/etna-team/etna/pull/618))
```
Comment (d-a-bunin) [inline CHANGELOG.md:49]: Why are there two new lines in changelog? This line isn't informative.

File: tests/test_datasets/test_utils.py [hunk 4]
```diff
@@ -1013,3 +1014,17 @@ def test_check_features_in_segments_ok(columns):
...
+@pytest.mark.parametrize("start, stop", ((0, 4), (4, -1), (-5, -1), (None, 6), (5, None), (None, None)))
+def test_slice_index_wide_dataframe_int_idx(df_aligned_datetime, start, stop):
+    res = _slice_index_wide_dataframe(df=df_aligned_datetime, start=start, stop=stop, label_indexing=False)
+    pd.testing.assert_frame_equal(res, df_aligned_datetime.iloc[start:stop])
+
+
+@pytest.mark.parametrize(
+    "start, stop", (("2020-01-01", "2020-01-04"), (None, "2020-01-10"), ("2020-01-09", None), (None, None))
+)
+def test_slice_index_wide_dataframe_label_idx(df_aligned_datetime, start, stop):
```
Comment (d-a-bunin) [inline tests/test_datasets/test_utils.py:1028]: There could also be label with int (for non-aligned).
But I don't think we should test that.

Comment (brsnw250) [reply tests/test_datasets/test_utils.py:1028]: This is just a dummy test to ensure that function indeed supports different types of indexing. Integer labeled indexing, tested with not aligned data.

**Review by d-a-bunin — APPROVED** [review]
